src/lightning/fabric/strategies/model_parallel.py

```diff
@@ -310,3 +358,206 @@ def __enter__(self) -> None:
+
+    return isinstance(module, Module) and any(isinstance(t, DTensor) for t in module.parameters())
+
+
+def _load_raw_module_state_from_path(path: Path, module: Module, world_size: int, strict: bool = True) -> None:
+    """Loads the state dict from a file path into the FSDP module."""
+    if not _is_full_checkpoint(path):
+        raise ValueError(
+            "Failed to load checkpoint directly into the model. The given path must be a single file containing the"
+            f" full state dict: {path}"
+        )
+    # Use `lazy_load` instead of `torch.load` here to avoid storing a copy of the full checkpoint per rank
+    _load_raw_module_state(state_dict=_lazy_load(path), module=module, world_size=world_size, strict=strict)
+
+
+def _load_raw_module_state(state_dict: Dict[str, Any], module: Module, world_size: int, strict: bool = True) -> None:
+    """Loads the state dict into the module by gathering all weights first and then and writing back to each shard."""
+    from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+
+    if _TORCH_GREATER_EQUAL_2_3 and _has_dtensor_modules(module):
+        _load_dtensor_state_dict(state_dict, module=module, strict=strict)
+    elif isinstance(module, FSDP):
+        with _get_full_state_dict_context(module, world_size=world_size, rank0_only=False):
+            module.load_state_dict(state_dict, strict=strict)
+    else:
+        module.load_state_dict(state_dict, strict=strict)
+
+
+def _load_dtensor_state_dict(state_dict: Dict[str, Any], module: Module, strict: bool = True) -> None:
+    # Note: There is currently no standard way in PyTorch to load a single-file checkpoint into a
+    # DTensor module and/or optimizer. We manually reshard and assign the tensors to the model, but don't
+    # support loading the optimizer state yet. Caveat: Custom `Module.load_state_dict()` methods are not supported
+    from torch.distributed._tensor import DTensor, distribute_tensor
+
+    module_keys = {name for name, _ in itertools.chain(module.named_parameters(), module.named_buffers())}
+    state_dict_keys = set(state_dict.keys())
+    if strict and module_keys != state_dict_keys:
+        raise KeyError(
+            "The state-dict keys in the model and checkpoint don't match. To disable strict loading,"
+            " set `strict=False`."
+        )
+
+    for name, param in itertools.chain(module.named_parameters(), module.named_buffers()):
+        if name not in state_dict:
+            # Strict loading already handled above
+            continue
+
+        loaded_tensor = state_dict[name]
+        if isinstance(loaded_tensor, _NotYetLoadedTensor):
+            loaded_tensor = loaded_tensor._load_tensor()
+        if isinstance(param, DTensor):
+            loaded_tensor = distribute_tensor(
+                tensor=loaded_tensor,
+                device_mesh=param.device_mesh,
+                placements=param.placements,
+            )
+        with torch.no_grad():
```

Comment:
This could be set as a decorator to the function. It might be useful to cover distribute_tensor too